xbps: support --rootdir and --repository

ssnailed opened the pull request:
##### SUMMARY
This MR adds support for the --rootdir and --repository flags of xbps-install, xbps-query and xbps-remove in the xbps module.

This makes it possible to provision new void linux installations ie. from a live image.

Fixes #9173

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
xbps

ssnailed commented:
As of now, this is untested. I will test this tomorrow.

felixfontein commented on plugins/modules/xbps.py:
```diff
@@ -67,6 +67,21 @@
+              package installation.
+              The URL can be a URL to a repository for
+              remote repositories or a path for local repositories.
+        aliases: [repositories]
```
Generally aliases should only be used if absolutely necessary. Here I would remove it, and potentially rename the option to `repositories`.

felixfontein reviewed (commented):
Thanks for your contribution!

ssnailed commented:
I got to this a bit later than expected, but I fixed some bugs and tested the new module parameters. 

During testing I noticed that xbps-install will create an interactive prompt asking if a new repo signing key should be accepted if it doesn't exist in the specified root directory. The solution to this would be copying the repo keys from the host system into the root directory. Is this something that should be handled by the module or explicitly handled by the user using a copy task? This could be accomplished like so:

```yaml
- name: Find repository keys to install into a new void system on a mounted partition
  ansible.builtin.find:
    path: /var/db/xbps/keys
    pattern: "*.plist"
  register: xbps_keys

- name: Copy repository keys to into a new void system on a mounted partition
  ansible.builtin.copy:
    remote_src: true
    src: "{{ item }}"
    dest: "/mnt/{{ item }}"
    owner: root
    group: root
    mode: "0644"
  loop: "{{ xbps_keys.files | map(attribute='path') }}"
  when: xbps_keys.matched > 0

- name: Install a new void system on a mounted partition
  community.general.xbps:
    name: base-system
    state: present
    repositories: https://repo-default.voidlinux.org/current
    root: /mnt
```

I've added the parameter `accept_pubkey` to allow the user to accept/deny the "Do you want to import this public key?" prompt, which could be insecure if the repository has been compromised.

I've also added an example for how to securely copy the signing keys from the host system.

Another thing that I noticed is that this module updates the package cache by default, even when simply removing a package. Is this intentional?

For reference, this is how I tested my changes:
```yaml
- name: Test xbps module
  hosts: dijkstra
  remote_user: root
  tasks:
    - name: Remove existing rootfs
      ansible.builtin.file:
        path: /tmp/test-rootfs
        state: absent

    - name: Create new rootfs
      ansible.builtin.file:
        path: /tmp/test-rootfs
        state: directory
        owner: root
        group: root
        mode: "0755"

    # This should fail because the repository keys don't exist in the rootfs
    - name: Install new rootfs
      community.general.xbps:
        name: base-system
        state: present
        repositories: https://repo-default.voidlinux.org/current
        root: /tmp/test-rootfs
      ignore_errors: true

    - name: Install new rootfs
      community.general.xbps:
        name: base-system
        state: present
        repositories: https://repo-default.voidlinux.org/current
        root: /tmp/test-rootfs
        accept_pubkey: true

    - name: Update package cache
      community.general.xbps:
        update_cache: true
        repositories: https://repo-default.voidlinux.org/current
        root: /tmp/test-rootfs
        accept_pubkey: true

    - name: Update package cache with custom repo
      community.general.xbps:
        update_cache: true
        repositories: https://repo-default.voidlinux.org/current
        accept_pubkey: true

    - name: Update package cache with custom repo and root
      community.general.xbps:
        update_cache: true
        repositories: https://repo-default.voidlinux.org/current
        root: /tmp/test-rootfs
        accept_pubkey: true

    - name: Install package
      community.general.xbps:
        name: dmenu-custom
        repositories:
          - https://repo-default.voidlinux.org/current
          - https://xbps.snaile.de
        root: /tmp/test-rootfs
        accept_pubkey: true
        state: present

    - name: Remove package
      community.general.xbps:
        name: dmenu-custom
        state: absent
        update_cache: false
        root: /tmp/test-rootfs
```

felixfontein commented on plugins/modules/xbps.py:
```diff
@@ -67,6 +67,26 @@
+        version_added: '10.1.0'
+    accept_pubkey:
+        description:
+            - Wether or not repository signing keys should be automatically accepted.
```
```suggestion
            - Whether or not repository signing keys should be automatically accepted.
```
(TIL that `wether` is also an [existing word](https://en.wikipedia.org/wiki/Wether), but definitely not what you meant ;) )

felixfontein commented:
I can't say much about your other questions, but:

> Another thing that I noticed is that this module updates the package cache by default, even when simply removing a package. Is this intentional?

No idea (I've never used the module or looked closer at it), but that doesn't sound like a good thing. It's probably best if this behavior is made configurable (with some boolean option?), with the default value equal to the current behavior. Then we can eventually deprecate the default value and switch it to a better default later.

ssnailed commented:
Currently package cache updates are controlled by the update_cache variable, which is true by default. At any rate, any changes would probably be outside the scope of this PR. 

With the typos fixed, I think this PR is ready to go. Thank you for your time!

the-maldridge commented on plugins/modules/xbps.py:
```diff
@@ -156,8 +213,15 @@ def query_package(module, xbps_path, name, state="present"):
-    rc, stdout, stderr = module.run_command(cmd, check_rc=False)
+    cmd = append_flags(module, xbps_path, cmd)
+    if module.params['accept_pubkey']:
+        stdin = "y\n"
```
The `-y` option does not approve key imports.  We've long talked about a non-interactive way to do key imports, but at the moment the accepted wisdom is to install the key plist into place if not passing a 'y' on stdin.

russoz replied:
Is there an issue lodged with the maintainers to have the `-y` work for that case, by any chance?

russoz replied:
Anyways, that's beside the point for this PR

ssnailed commented on plugins/modules/xbps.py:
```diff
@@ -67,6 +67,26 @@
+    root:
+        description:
+            - The full path for the target root directory.
+        type: str
```
I wasn't aware that's an option, that seems like they way to go.

russoz commented on plugins/modules/xbps.py:
```diff
@@ -308,6 +377,9 @@ def main():
             upgrade=dict(default=False, type='bool'),
             update_cache=dict(default=True, type='bool'),
             upgrade_xbps=dict(default=True, type='bool'),
+            root=dict(type='str'),
```
If you change it to path, that change should be applied to both this line and the docs above.

russoz commented on plugins/modules/xbps.py:
```diff
@@ -67,6 +67,26 @@
+        description:
+            - The full path for the target root directory.
+        type: path
+        version_added: '10.1.0'
```
10.1.0 has been released, please update (here and in the other placea in the files):
```suggestion
        version_added: '10.2.0'
```

russoz reviewed (approved):
LGTM

felixfontein commented:
@ssnailed thanks for your contribution!
@russoz @the-maldridge thanks for reviewing!
